src/channel.rs

```diff
@@ -0,0 +1,284 @@
+    }
+}
+
+impl Evented for ReceiverCtl {
+    fn register(&self, poll: &Poll, token: Token, interest: EventSet, opts: PollOpt) -> io::Result<()> {
+        if self.registration.is_some() {
+            return Err(io::Error::new(io::ErrorKind::Other, "receiver already registered"));
+        }
+
+        let (registration, set_readiness) = Registration::new(poll, token, interest, opts);
+
+
+        if self.inner.pending.load(Ordering::Relaxed) > 0 {
+            // TODO: Don't drop readiness
+            let _ = set_readiness.set_readiness(EventSet::readable());
+        }
+
+        self.registration.set(registration).ok().expect("unexpected state encountered");
+        self.inner.set_readiness.set(set_readiness).ok().expect("unexpected state encountered");
+
+        Ok(())
+    }
+
+    fn reregister(&self, poll: &Poll, token: Token, interest: EventSet, opts: PollOpt) -> io::Result<()> {
+        match self.registration.as_ref() {
+            Some(registration) => registration.update(poll, token, interest, opts),
+            None => Err(io::Error::new(io::ErrorKind::Other, "receiver not registered")),
+        }
+    }
+
+    fn deregister(&self, poll: &Poll) -> io::Result<()> {
+        match self.registration.as_ref() {
+            Some(registration) => registration.deregister(poll),
+            None => Err(io::Error::new(io::ErrorKind::Other, "receiver not registered")),
+        }
+    }
+}
+
+impl<T> Sender<T> {
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        self.tx.send(t).and_then(|_| {
+            try!(self.ctl.inc());
+            Ok(())
+        })
+    }
+
+    pub fn try_send(&self, t: T) -> Result<(), TrySendError<T>> {
+        self.tx.try_send(t).and_then(|_| {
+            try!(self.ctl.inc());
+            Ok(())
+        })
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Sender<T> {
+        Sender {
+            tx: self.tx.clone(),
+            ctl: self.ctl.clone(),
+        }
+    }
+}
+
+impl<T> StdSender<T> {
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        match *self {
+            StdSender::Bounded(ref tx) => tx.send(t).map_err(SendError::from),
+            StdSender::Unbounded(ref tx) => tx.send(t).map_err(SendError::from),
+        }
+    }
+
+    pub fn try_send(&self, t: T) -> Result<(), TrySendError<T>> {
+        match *self {
+            StdSender::Bounded(ref tx) => tx.try_send(t).map_err(TrySendError::from),
+            StdSender::Unbounded(ref tx) => tx.send(t).map_err(TrySendError::from),
+        }
+    }
+}
+
+impl<T> Clone for StdSender<T> {
+    fn clone(&self) -> StdSender<T> {
+        match *self {
+            StdSender::Bounded(ref v) => StdSender::Bounded(v.clone()),
+            StdSender::Unbounded(ref v) => StdSender::Unbounded(v.clone()),
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    pub fn try_recv(&self) -> Result<T, mpsc::TryRecvError> {
+        // TODO: When there is no value to recv, unset readiness
```

Comment:
Yes, `dec` will unset readiness. I think I was suggesting something else... I also think the TODO here is out of date. I will remove.